src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/v2/NpgStateService.kt

```diff
@@ -0,0 +1,125 @@
+
+import com.azure.core.util.BinaryData
+import com.azure.storage.queue.QueueAsyncClient
+import it.pagopa.ecommerce.commons.client.NpgClient
+import it.pagopa.ecommerce.commons.documents.v2.TransactionAuthorizationRequestedEvent
+import it.pagopa.ecommerce.commons.domain.v2.pojos.BaseTransaction
+import it.pagopa.ecommerce.commons.domain.v2.pojos.BaseTransactionWithPaymentToken
+import it.pagopa.ecommerce.commons.exceptions.NpgResponseException
+import it.pagopa.ecommerce.commons.generated.npg.v1.dto.StateResponseDto
+import it.pagopa.ecommerce.commons.queues.QueueEvent
+import it.pagopa.ecommerce.commons.queues.TracingInfo
+import it.pagopa.ecommerce.commons.utils.NpgPspApiKeysConfig
+import it.pagopa.ecommerce.eventdispatcher.exceptions.BadGatewayException
+import it.pagopa.ecommerce.eventdispatcher.exceptions.GetStateException
+import it.pagopa.ecommerce.eventdispatcher.exceptions.NoRetryAttemptsLeftException
+import it.pagopa.ecommerce.eventdispatcher.exceptions.TooLateRetryAttemptException
+import it.pagopa.ecommerce.eventdispatcher.queues.v2.QueueCommonsLogger.logger
+import java.time.Duration
+import java.time.Instant
+import java.util.*
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+import reactor.core.publisher.Mono
+import reactor.kotlin.core.publisher.switchIfEmpty
+
+@Component
+class NpgStateService(
+  @Autowired private val npgClient: NpgClient,
+  @Autowired private val npgCardsPspApiKey: NpgPspApiKeysConfig,
+  @Value("\${transactionAuthorizationRequested.paymentTokenValidityTimeOffset}")
+  private val paymentTokenValidityTimeOffset: Int,
+  @Autowired private val authRequestedQueueAsyncClient: QueueAsyncClient,
+  @Value("\${transactionAuthorizationRequested.eventOffsetSeconds}") private val retryOffset: Int,
+  @Value("\${transactionAuthorizationRequested.maxAttempts}") private val maxAttempts: Int,
+  @Value("\${azurestorage.queues.transientQueues.ttlSeconds}")
+  private val transientQueuesTTLSeconds: Int,
+) {
+
+  fun getStateNpg(transactionId: UUID, sessionId: String, pspId: String): Mono<StateResponseDto> {
+    return npgCardsPspApiKey[pspId].fold(
+      { ex -> Mono.error(ex) },
+      { apiKey ->
+        npgClient.getState(UUID.randomUUID(), sessionId, apiKey).onErrorMap(
+          NpgResponseException::class.java) { exception: NpgResponseException ->
+          val responseStatusCode = exception.statusCode
+          responseStatusCode
+            .map {
+              val errorCodeReason = "Received HTTP error code from NPG: $it"
+              if (it.is5xxServerError) {
+                BadGatewayException(errorCodeReason)
+              } else {
+                GetStateException(transactionId, errorCodeReason)
+              }
+            }
+            .orElse(GetStateException(transactionId, "Unknown NPG HTTP response code"))
+        }
+      })
+  }
+
+  fun enqueueRetryEvent(
```

Comment:
Why re-implementing those methods? you can make this class override [this](https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/blob/main/src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/EventReceiverService.kt)
overriding the abstract methods as done [here](https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/blob/main/src/main/kotlin/it/pagopa/ecommerce/eventdispatcher/services/eventretry/v2/ClosureRetryService.kt) for example

Reply:
To be addressed in future task

Reply:
fixed with https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/216/commits/f0e2f01e336eb976065a989d489d3647a1331615 making this class extend the proper super class that already handles retry event :) retry event logic have been short-circuited in order to throw exception waiting to retry event creation in commons library as per the future task

Reply:
furthermore npg state logic have been splitted from retrieve authorization logic with https://github.com/pagopa/pagopa-ecommerce-event-dispatcher-service/pull/216/commits/b63edf958b50dd1015d74b9577f8f91e2c3fcbb9#diff-bfe3aa2e68cfd3a13941ca06afec3c6d4a23dccaa60f86ddfc20dcd44deccd6e in order to differentiate event retry service logic from Npg logic